### ISSUE TITLE
sbt reports success when forked test suite fails with an exception

### ISSUE THREAD

**Issue by smarter** [sbt/sbt#2442]
See https://scala-ci.typesafe.com/job/dotty-master-validate-junit/1000/console for an example:

``` scala
Internal error when running tests: java.lang.OutOfMemoryError: Java heap space
Exception in thread "Thread-7" java.io.EOFException
    at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2601)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1319)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
    at java.io.ObjectInputStream.readObject(ObjectInputStream.java:371)
    at sbt.React.react(ForkTests.scala:114)
    at sbt.ForkTests$$anonfun$mainTestTask$1$Acceptor$2$.run(ForkTests.scala:74)
    at java.lang.Thread.run(Thread.java:745)
[info] Passed: Total 8, Failed 0, Errors 0, Passed 8
[success] Total time: 1721 s, completed Feb 4, 2016 8:53:52 PM
```

**Comment by SethTisue** [comment]
this makes me a bit nervous about switching scala/scala PR validation from ant to sbt, since sbt will be running some tests forked. did you do something special to work around this for Dotty?

**Comment by smarter** [comment]
No, I just fixed the underlying issue that led to all these OutOfMemoryErrors. This looks like something that can only really be fixed in sbt itself.

**Comment by smarter** [comment]
Note that this should be pretty rare: it happens when the test suite itself fails with an exception, not when a test fails with an exception.

**Comment by eed3si9n** [comment]
There was similar report with ScalaTest sbt/sbt#1890 / scalatest/scalatest#432.
Not sure what they did (maybe scalatest/scalatest@a3cf999053108f9fe63ee984f0221c55fdf76dcc) but ScalaTest 2.2.6 apparently fixed it.

**Comment by smarter** [comment]
SBT does not deal with exceptions in threads correctly, see: https://github.com/sbt/sbt/blob/a92e50929c56a0eec160b3c7fc1cddf7512309ff/main/actions/src/main/scala/sbt/ForkTests.scala#L39-L79
`Acceptor` is an instance of `Runnable`, in its `run` method you do not catch any exception, which mean that if one occurs, the thread dies, but nothing will be reported to the caller because you're just using `Thread#run` and `Thread#join` which don't report anything about how a Thread died.
I don't know much about Java concurrency but I guess you could use https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.UncaughtExceptionHandler.html

**Comment by parkan** [comment]
Seeing this with copycat + sbt as well

https://circleci.com/gh/mediachain/mediachain/48

Definitely doesn't seem like the correct behavior

**Comment by kevinmeredith** [comment]
> but ScalaTest 2.2.6 apparently fixed it.

Does this [gist](https://gist.github.com/kevinmeredith/706cb44d2e984ad838aad7ded4c048a6) indicate its success, @eed3si9n?

**Comment by eed3si9n** [comment]
@kevinmeredith Reading this issue over today, I now think sbt/sbt#1890 is unrelated.

**Comment by dwijnand** [comment]
Fixed in #2730 by @eed3si9n
